nomos-cli/src/da/disseminate.rs

```diff
@@ -0,0 +1,251 @@
+use clap::{Args, ValueEnum};
+use full_replication::{AbsoluteNumber, Attestation, Certificate, FullReplication};
+use futures::StreamExt;
+use nomos_core::{da::DaProtocol, wire};
+use nomos_da::network::{adapters::libp2p::Libp2pAdapter, NetworkAdapter};
+use nomos_network::{backends::libp2p::Libp2p, NetworkService};
+use overwatch_derive::*;
+use overwatch_rs::{
+    services::{
+        handle::{ServiceHandle, ServiceStateHandle},
+        relay::NoMessage,
+        state::*,
+        ServiceCore, ServiceData, ServiceId,
+    },
+    DynError,
+};
+use reqwest::{Client, Url};
+use serde::Serialize;
+use std::{
+    path::PathBuf,
+    sync::{mpsc::Sender, Arc},
+    time::Duration,
+};
+use tokio::sync::{mpsc::UnboundedReceiver, Mutex};
+
+const NODE_CERT_PATH: &str = "mempool-da/add";
+
+pub async fn disseminate_and_wait<D, B, N, A, C>(
+    mut da: D,
+    data: Box<[u8]>,
+    adapter: N,
+    status_updates: Sender<Status>,
+    node_addr: Option<&Url>,
+    output: Option<&PathBuf>,
+) -> Result<(), Box<dyn std::error::Error>>
+where
+    D: DaProtocol<Blob = B, Attestation = A, Certificate = C>,
+    N: NetworkAdapter<Blob = B, Attestation = A> + Send + Sync,
+    C: Serialize,
+{
+    // 1) Building blob
+    status_updates.send(Status::Encoding)?;
+    let blobs = da.encode(data);
+
+    // 2) Send blob to network
+    status_updates.send(Status::Disseminating)?;
+    futures::future::try_join_all(blobs.into_iter().map(|blob| adapter.send_blob(blob)))
```

Comment:
This got me thinking. If we just need a threshold of responses would it be also interesting to consider that we may fail to disseminate a few of them? 🤔